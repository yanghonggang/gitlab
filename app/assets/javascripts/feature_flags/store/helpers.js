import { isEmpty, uniqueId, isString } from 'lodash';
import {
  ROLLOUT_STRATEGY_ALL_USERS,
  ROLLOUT_STRATEGY_PERCENT_ROLLOUT,
  ROLLOUT_STRATEGY_USER_ID,
  ROLLOUT_STRATEGY_GITLAB_USER_LIST,
  INTERNAL_ID_PREFIX,
  DEFAULT_PERCENT_ROLLOUT,
  PERCENT_ROLLOUT_GROUP_ID,
  fetchPercentageParams,
  fetchUserIdParams,
  LEGACY_FLAG,
} from '../constants';

/**
 * Converts raw scope objects fetched from the API into an array of scope
 * objects that is easier/nicer to bind to in Vue.
 * @param {Array} scopesFromRails An array of scope objects fetched from the API
 */
export const mapToScopesViewModel = scopesFromRails =>
  (scopesFromRails || []).map(s => {
    const percentStrategy = (s.strategies || []).find(
      strat => strat.name === ROLLOUT_STRATEGY_PERCENT_ROLLOUT,
    );

    const rolloutPercentage = fetchPercentageParams(percentStrategy) || DEFAULT_PERCENT_ROLLOUT;

    const userStrategy = (s.strategies || []).find(
      strat => strat.name === ROLLOUT_STRATEGY_USER_ID,
    );

    const rolloutStrategy =
      (percentStrategy && percentStrategy.name) ||
      (userStrategy && userStrategy.name) ||
      ROLLOUT_STRATEGY_ALL_USERS;

    const rolloutUserIds = (fetchUserIdParams(userStrategy) || '')
      .split(',')
      .filter(id => id)
      .join(', ');

    return {
      id: s.id,
      environmentScope: s.environment_scope,
      active: Boolean(s.active),
      canUpdate: Boolean(s.can_update),
      protected: Boolean(s.protected),
      rolloutStrategy,
      rolloutPercentage,
      rolloutUserIds,

      // eslint-disable-next-line no-underscore-dangle
      shouldBeDestroyed: Boolean(s._destroy),
      shouldIncludeUserIds: rolloutUserIds.length > 0 && percentStrategy !== null,
    };
  });
/**
 * Converts the parameters emitted by the Vue component into
 * the shape that the Rails API expects.
 * @param {Array} scopesFromVue An array of scope objects from the Vue component
 */
export const mapFromScopesViewModel = params => {
  const scopes = (params.scopes || []).map(s => {
    const parameters = {};
    if (s.rolloutStrategy === ROLLOUT_STRATEGY_PERCENT_ROLLOUT) {
      parameters.groupId = PERCENT_ROLLOUT_GROUP_ID;
      parameters.percentage = s.rolloutPercentage;
    } else if (s.rolloutStrategy === ROLLOUT_STRATEGY_USER_ID) {
      parameters.userIds = (s.rolloutUserIds || '').replace(/, /g, ',');
    }

    const userIdParameters = {};

    if (s.shouldIncludeUserIds && s.rolloutStrategy !== ROLLOUT_STRATEGY_USER_ID) {
      userIdParameters.userIds = (s.rolloutUserIds || '').replace(/, /g, ',');
    }

    // Strip out any internal IDs
    const id = isString(s.id) && s.id.startsWith(INTERNAL_ID_PREFIX) ? undefined : s.id;

    const strategies = [
      {
        name: s.rolloutStrategy,
        parameters,
      },
    ];

    if (!isEmpty(userIdParameters)) {
      strategies.push({ name: ROLLOUT_STRATEGY_USER_ID, parameters: userIdParameters });
    }

    return {
      id,
      environment_scope: s.environmentScope,
      active: s.active,
      can_update: s.canUpdate,
      protected: s.protected,
      _destroy: s.shouldBeDestroyed,
      strategies,
    };
  });

  const model = {
    operations_feature_flag: {
      name: params.name,
      description: params.description,
      active: params.active,
      scopes_attributes: scopes,
      version: LEGACY_FLAG,
    },
  };

  return model;
};

/**
 * Creates a new feature flag environment scope object for use
 * in a Vue component.  An optional parameter can be passed to
 * override the property values that are created by default.
 *
 * @param {Object} overrides An optional object whose
 * property values will be used to override the default values.
 *
 */
export const createNewEnvironmentScope = (overrides = {}, featureFlagPermissions = false) => {
  const defaultScope = {
    environmentScope: '',
    active: false,
    id: uniqueId(INTERNAL_ID_PREFIX),
    rolloutStrategy: ROLLOUT_STRATEGY_ALL_USERS,
    rolloutPercentage: DEFAULT_PERCENT_ROLLOUT,
    rolloutUserIds: '',
  };

  const newScope = {
    ...defaultScope,
    ...overrides,
  };

  if (featureFlagPermissions) {
    newScope.canUpdate = true;
    newScope.protected = false;
  }

  return newScope;
};

const mapStrategyScopesToRails = scopes =>
  scopes.length === 0
    ? [{ environment_scope: '*' }]
    : scopes.map(s => ({
        id: s.id,
        _destroy: s.shouldBeDestroyed,
        environment_scope: s.environmentScope,
      }));

const mapStrategyScopesToView = scopes =>
  scopes.map(s => ({
    id: s.id,
    // eslint-disable-next-line no-underscore-dangle
    shouldBeDestroyed: Boolean(s._destroy),
    environmentScope: s.environment_scope,
  }));

const mapStrategiesParametersToViewModel = params => {
  if (params.userIds) {
    return { ...params, userIds: params.userIds.split(',').join(', ') };
  }
  return params;
};

export const mapStrategiesToViewModel = strategiesFromRails =>
  (strategiesFromRails || []).map(s => ({
    id: s.id,
    name: s.name,
    parameters: mapStrategiesParametersToViewModel(s.parameters),
    userList: s.user_list,
    // eslint-disable-next-line no-underscore-dangle
    shouldBeDestroyed: Boolean(s._destroy),
    scopes: mapStrategyScopesToView(s.scopes),
  }));

const mapStrategiesParametersToRails = params => {
  if (params.userIds) {
    return { ...params, userIds: params.userIds.replace(/\s*,\s*/g, ',') };
  }
  return params;
};

const mapStrategyToRails = strategy => {
  const mappedStrategy = {
    id: strategy.id,
    name: strategy.name,
    _destroy: strategy.shouldBeDestroyed,
    scopes_attributes: mapStrategyScopesToRails(strategy.scopes || []),
    parameters: mapStrategiesParametersToRails(strategy.parameters),
  };

  if (strategy.name === ROLLOUT_STRATEGY_GITLAB_USER_LIST) {
    mappedStrategy.user_list_id = strategy.userList.id;
  }
  return mappedStrategy;
};

export const mapStrategiesToRails = params => ({
  operations_feature_flag: {
    name: params.name,
    description: params.description,
    version: params.version,
    active: params.active,
    strategies_attributes: (params.strategies || []).map(mapStrategyToRails),
  },
});
