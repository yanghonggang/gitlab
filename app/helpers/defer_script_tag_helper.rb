# frozen_string_literal: true

module DeferScriptTagHelper
  # Override the default ActionView `javascript_include_tag` helper to support page specific deferred loading.
  # PLEASE NOTE: `defer` is also critical so that we don't run JavaScript entrypoints before the DOM is ready.
  # Please see https://gitlab.com/groups/gitlab-org/-/epics/4538#note_432159769.
  def javascript_include_tag(*sources)
    super(*sources, defer: true)
  end
end
