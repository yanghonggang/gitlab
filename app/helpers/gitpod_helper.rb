# frozen_string_literal: true

module GitpodHelper
  def gitpod_enable_description
    s_('Enable %{linkStart}Gitpod%{linkEnd} integration to launch a development environment in your browser directly from GitLab.')
  end
end
