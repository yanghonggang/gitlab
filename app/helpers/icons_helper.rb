# frozen_string_literal: true

require 'json'

module IconsHelper
  extend self
  include FontAwesome::Rails::IconHelper

  DEFAULT_ICON_SIZE = 16

  # Creates an icon tag given icon name(s) and possible icon modifiers.
  #
  # Right now this method simply delegates directly to `fa_icon` from the
  # font-awesome-rails gem, but should we ever use a different icon pack in the
  # future we won't have to change hundreds of method calls.
  def icon(names, options = {})
    if (options.keys & %w[aria-hidden aria-label data-hidden]).empty?
      # Add 'aria-hidden' and 'data-hidden' if they are not set in options.
      options['aria-hidden'] = true
      options['data-hidden'] = true
    end

    options.include?(:base) ? fa_stacked_icon(names, options) : fa_icon(names, options)
  end

  def custom_icon(icon_name, size: DEFAULT_ICON_SIZE)
    # We can't simply do the below, because there are some .erb SVGs.
    #  File.read(Rails.root.join("app/views/shared/icons/_#{icon_name}.svg")).html_safe
    render "shared/icons/#{icon_name}.svg", size: size
  end

  def sprite_icon_path
    @sprite_icon_path ||= begin
      # SVG Sprites currently don't work across domains, so in the case of a CDN
      # we have to set the current path deliberately to prevent addition of asset_host
      sprite_base_url = Gitlab.config.gitlab.url if ActionController::Base.asset_host
      ActionController::Base.helpers.image_path('icons.svg', host: sprite_base_url)
    end
  end

  def sprite_file_icons_path
    # SVG Sprites currently don't work across domains, so in the case of a CDN
    # we have to set the current path deliberately to prevent addition of asset_host
    sprite_base_url = Gitlab.config.gitlab.url if ActionController::Base.asset_host
    ActionController::Base.helpers.image_path('file_icons.svg', host: sprite_base_url)
  end

  def sprite_icon(icon_name, size: DEFAULT_ICON_SIZE, css_class: nil)
    if known_sprites&.exclude?(icon_name)
      exception = ArgumentError.new("#{icon_name} is not a known icon in @gitlab-org/gitlab-svg")
      Gitlab::ErrorTracking.track_and_raise_for_dev_exception(exception)
    end

    css_classes = []
    css_classes << "s#{size}" if size
    css_classes << "#{css_class}" unless css_class.blank?

    content_tag(
      :svg,
      content_tag(:use, '', { 'xlink:href' => "#{sprite_icon_path}##{icon_name}" } ),
      class: css_classes.empty? ? nil : css_classes.join(' '),
      data: { testid: "#{icon_name}-icon" }
    )
  end

  def loading_icon(container: false, color: 'orange', size: 'sm', css_class: nil)
    css_classes = ['gl-spinner', "gl-spinner-#{color}", "gl-spinner-#{size}"]
    css_classes << "#{css_class}" unless css_class.blank?

    spinner = content_tag(:span, "", { class: css_classes.join(' '), aria: { label: _('Loading') } })

    container == true ? content_tag(:div, spinner, { class: 'gl-spinner-container' }) : spinner
  end

  def external_snippet_icon(name)
    content_tag(:span, "", class: "gl-snippet-icon gl-snippet-icon-#{name}")
  end

  def audit_icon(name, css_class: nil)
    case name
    when "standard"
      name = "key"
    when "two-factor"
      name = "key"
    when "google_oauth2"
      name = "google"
    end

    sprite_icon(name, css_class: css_class)
  end

  def boolean_to_icon(value)
    if value
      sprite_icon('check', css_class: 'cgreen')
    else
      sprite_icon('power', css_class: 'clgray')
    end
  end

  def visibility_level_icon(level, options: {})
    name =
      case level
      when Gitlab::VisibilityLevel::PRIVATE
        'lock'
      when Gitlab::VisibilityLevel::INTERNAL
        'shield'
      else # Gitlab::VisibilityLevel::PUBLIC
        'earth'
      end

    css_class = options.delete(:class)

    sprite_icon(name, size: DEFAULT_ICON_SIZE, css_class: css_class)
  end

  def file_type_icon_class(type, mode, name)
    if type == 'folder'
      icon_class = 'folder-o'
    elsif type == 'archive'
      icon_class = 'archive'
    elsif mode == '120000'
      icon_class = 'share'
    else
      # Guess which icon to choose based on file extension.
      # If you think a file extension is missing, feel free to add it on PR

      case File.extname(name).downcase
      when '.pdf'
        icon_class = 'document'
      when '.jpg', '.jpeg', '.jif', '.jfif',
          '.jp2', '.jpx', '.j2k', '.j2c',
          '.apng', '.png', '.gif', '.tif', '.tiff',
          '.svg', '.ico', '.bmp', '.webp'
        icon_class = 'doc-image'
      when '.zip', '.zipx', '.tar', '.gz', '.gzip', '.tgz', '.bz', '.bzip',
          '.bz2', '.bzip2', '.car', '.tbz', '.xz', 'txz', '.rar', '.7z',
          '.lz', '.lzma', '.tlz'
        icon_class = 'doc-compressed'
      when '.mp3', '.wma', '.ogg', '.oga', '.wav', '.flac', '.aac', '.3ga',
          '.ac3', '.midi', '.m4a', '.ape', '.mpa'
        icon_class = 'volume-up'
      when '.mp4', '.m4p', '.m4v',
          '.mpg', '.mp2', '.mpeg', '.mpe', '.mpv',
          '.mpg', '.mpeg', '.m2v', '.m2ts',
          '.avi', '.mkv', '.flv', '.ogv', '.mov',
          '.3gp', '.3g2'
        icon_class = 'live-preview'
      when '.doc', '.dot', '.docx', '.docm', '.dotx', '.dotm', '.docb',
          '.odt', '.ott', '.uot', '.rtf'
        icon_class = 'doc-text'
      when '.xls', '.xlt', '.xlm', '.xlsx', '.xlsm', '.xltx', '.xltm',
          '.xlsb', '.xla', '.xlam', '.xll', '.xlw', '.ots', '.ods', '.uos'
        icon_class = 'document'
      when '.ppt', '.pot', '.pps', '.pptx', '.pptm', '.potx', '.potm',
          '.ppam', '.ppsx', '.ppsm', '.sldx', '.sldm', '.odp', '.otp', '.uop'
        icon_class = 'doc-chart'
      else
        icon_class = 'doc-text'
      end
    end

    icon_class
  end

  private

  def known_sprites
    return if Rails.env.production?

    @known_sprites ||= Gitlab::Json.parse(File.read(Rails.root.join('node_modules/@gitlab/svgs/dist/icons.json')))['icons']
  end
end
