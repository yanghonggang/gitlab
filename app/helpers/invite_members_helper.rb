# frozen_string_literal: true

module InviteMembersHelper
  include Gitlab::Utils::StrongMemoize

  def invite_members_allowed?(group)
    Feature.enabled?(:invite_members_group_modal, group) && can?(current_user, :admin_group_member, group)
  end

  def directly_invite_members?
    strong_memoize(:directly_invite_members) do
      experiment_enabled?(:invite_members_version_a) && can_import_members?
    end
  end

  def indirectly_invite_members?
    strong_memoize(:indirectly_invite_members) do
      experiment_enabled?(:invite_members_version_b) && !can_import_members?
    end
  end

  def invite_group_members?(group)
    experiment_enabled?(:invite_members_empty_group_version_a) && Ability.allowed?(current_user, :admin_group_member, group)
  end
end
