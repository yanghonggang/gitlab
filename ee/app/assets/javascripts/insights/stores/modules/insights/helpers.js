import { CHART_TYPES } from 'ee/insights/constants';
import { __ } from '~/locale';

const getAxisTitle = label => {
  switch (label) {
    case 'day':
      return __('Days');
    case 'week':
      return __('Weeks');
    case 'month':
      return __('Months');
    case 'issue':
      return __('Issues');
    case 'merge_request':
      return __('Merge requests');
    default:
      return '';
  }
};

export const transformChartDataForGlCharts = (
  { type, query: { group_by, issuable_type } },
  { labels, datasets },
) => {
  const formattedData = {
    xAxisTitle: getAxisTitle(group_by),
    yAxisTitle: getAxisTitle(issuable_type),
    labels,
    datasets: [],
    seriesNames: [],
  };

  switch (type) {
    case CHART_TYPES.BAR:
      formattedData.datasets = [
        {
          name: 'all',
          data: labels.map((label, i) => [label, datasets[0].data[i]]),
        },
      ];
      break;
    case CHART_TYPES.STACKED_BAR:
      formattedData.datasets.push(
        ...datasets.map(dataset => ({
          name: dataset.label,
          data: dataset.data,
        })),
      );
      break;
    case CHART_TYPES.LINE:
      formattedData.datasets.push(
        ...datasets.map(dataset => ({
          name: dataset.label,
          data: labels.map((label, i) => [label, dataset.data[i]]),
        })),
      );

      break;
    default:
      formattedData.datasets = { all: labels.map((label, i) => [label, datasets[0].data[i]]) };
  }
  return formattedData;
};

export default {
  transformChartDataForGlCharts,
};
