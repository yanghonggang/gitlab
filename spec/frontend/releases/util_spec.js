import { cloneDeep } from 'lodash';
import { getJSONFixture } from 'helpers/fixtures';
import {
  releaseToApiJson,
  apiJsonToRelease,
  convertGraphQLRelease,
  convertAllReleasesGraphQLResponse,
  convertOneReleaseGraphQLResponse,
} from '~/releases/util';

const originalAllReleasesQueryResponse = getJSONFixture(
  'graphql/releases/queries/all_releases.query.graphql.json',
);
const originalOneReleaseQueryResponse = getJSONFixture(
  'graphql/releases/queries/one_release.query.graphql.json',
);

describe('releases/util.js', () => {
  describe('releaseToApiJson', () => {
    it('converts a release JavaScript object into JSON that the Release API can accept', () => {
      const release = {
        tagName: 'tag-name',
        name: 'Release name',
        description: 'Release description',
        milestones: ['13.2', '13.3'],
        assets: {
          links: [{ url: 'https://gitlab.example.com/link', linkType: 'other' }],
        },
      };

      const expectedJson = {
        tag_name: 'tag-name',
        ref: null,
        name: 'Release name',
        description: 'Release description',
        milestones: ['13.2', '13.3'],
        assets: {
          links: [{ url: 'https://gitlab.example.com/link', link_type: 'other' }],
        },
      };

      expect(releaseToApiJson(release)).toEqual(expectedJson);
    });

    describe('when createFrom is provided', () => {
      it('adds the provided createFrom ref to the JSON as a "ref" property', () => {
        const createFrom = 'main';

        const release = {};

        const expectedJson = {
          ref: createFrom,
        };

        expect(releaseToApiJson(release, createFrom)).toMatchObject(expectedJson);
      });
    });

    describe('release.name', () => {
      it.each`
        input                 | output
        ${null}               | ${null}
        ${''}                 | ${null}
        ${' \t\n\r\n'}        | ${null}
        ${'  Release name  '} | ${'Release name'}
      `('converts a name like `$input` to `$output`', ({ input, output }) => {
        const release = { name: input };

        const expectedJson = {
          name: output,
        };

        expect(releaseToApiJson(release)).toMatchObject(expectedJson);
      });
    });
  });

  describe('apiJsonToRelease', () => {
    it('converts JSON received from the Release API into an object usable by the Vue application', () => {
      const json = {
        tag_name: 'tag-name',
        assets: {
          links: [
            {
              link_type: 'other',
            },
          ],
        },
      };

      const expectedRelease = {
        tagName: 'tag-name',
        assets: {
          links: [
            {
              linkType: 'other',
            },
          ],
        },
        milestones: [],
      };

      expect(apiJsonToRelease(json)).toEqual(expectedRelease);
    });
  });

  describe('convertGraphQLRelease', () => {
    let releaseFromResponse;
    let convertedRelease;

    beforeEach(() => {
      releaseFromResponse = cloneDeep(originalOneReleaseQueryResponse).data.project.release;
      convertedRelease = convertGraphQLRelease(releaseFromResponse);
    });

    describe('assets', () => {
      it("handles asset links that don't have a linkType", () => {
        expect(convertedRelease.assets.links[0].linkType).not.toBeUndefined();

        delete releaseFromResponse.assets.links.nodes[0].linkType;

        convertedRelease = convertGraphQLRelease(releaseFromResponse);

        expect(convertedRelease.assets.links[0].linkType).toBeUndefined();
      });
    });

    describe('_links', () => {
      it("handles releases that don't have any links", () => {
        expect(convertedRelease._links.selfUrl).not.toBeUndefined();

        delete releaseFromResponse.links;

        convertedRelease = convertGraphQLRelease(releaseFromResponse);

        expect(convertedRelease._links.selfUrl).toBeUndefined();
      });
    });

    describe('commit', () => {
      it("handles releases that don't have any commit info", () => {
        expect(convertedRelease.commit).not.toBeUndefined();

        delete releaseFromResponse.commit;

        convertedRelease = convertGraphQLRelease(releaseFromResponse);

        expect(convertedRelease.commit).toBeUndefined();
      });
    });
  });

  describe('convertAllReleasesGraphQLResponse', () => {
    it('matches snapshot', () => {
      expect(convertAllReleasesGraphQLResponse(originalAllReleasesQueryResponse)).toMatchSnapshot();
    });
  });

  describe('convertOneReleaseGraphQLResponse', () => {
    it('matches snapshot', () => {
      expect(convertOneReleaseGraphQLResponse(originalOneReleaseQueryResponse)).toMatchSnapshot();
    });
  });
});
