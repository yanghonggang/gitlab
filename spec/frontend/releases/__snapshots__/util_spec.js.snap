// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`releases/util.js convertAllReleasesGraphQLResponse matches snapshot 1`] = `
Object {
  "data": Array [
    Object {
      "_links": Object {
        "closedIssuesUrl": "http://localhost/releases-namespace/releases-project/-/issues?release_tag=v1.1&scope=all&state=closed",
        "closedMergeRequestsUrl": "http://localhost/releases-namespace/releases-project/-/merge_requests?release_tag=v1.1&scope=all&state=closed",
        "editUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/edit",
        "mergedMergeRequestsUrl": "http://localhost/releases-namespace/releases-project/-/merge_requests?release_tag=v1.1&scope=all&state=merged",
        "openedIssuesUrl": "http://localhost/releases-namespace/releases-project/-/issues?release_tag=v1.1&scope=all&state=opened",
        "openedMergeRequestsUrl": "http://localhost/releases-namespace/releases-project/-/merge_requests?release_tag=v1.1&scope=all&state=opened",
        "self": "http://localhost/releases-namespace/releases-project/-/releases/v1.1",
        "selfUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1",
      },
      "assets": Object {
        "count": 8,
        "links": Array [
          Object {
            "directAssetUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/downloads/binaries/awesome-app-3",
            "external": true,
            "id": "gid://gitlab/Releases::Link/13",
            "linkType": "image",
            "name": "Image",
            "url": "https://example.com/image",
          },
          Object {
            "directAssetUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/downloads/binaries/awesome-app-2",
            "external": true,
            "id": "gid://gitlab/Releases::Link/12",
            "linkType": "package",
            "name": "Package",
            "url": "https://example.com/package",
          },
          Object {
            "directAssetUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/downloads/binaries/awesome-app-1",
            "external": false,
            "id": "gid://gitlab/Releases::Link/11",
            "linkType": "runbook",
            "name": "Runbook",
            "url": "http://localhost/releases-namespace/releases-project/runbook",
          },
          Object {
            "directAssetUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/downloads/binaries/linux-amd64",
            "external": true,
            "id": "gid://gitlab/Releases::Link/10",
            "linkType": "other",
            "name": "linux-amd64 binaries",
            "url": "https://downloads.example.com/bin/gitlab-linux-amd64",
          },
        ],
        "sources": Array [
          Object {
            "format": "zip",
            "url": "http://localhost/releases-namespace/releases-project/-/archive/v1.1/releases-project-v1.1.zip",
          },
          Object {
            "format": "tar.gz",
            "url": "http://localhost/releases-namespace/releases-project/-/archive/v1.1/releases-project-v1.1.tar.gz",
          },
          Object {
            "format": "tar.bz2",
            "url": "http://localhost/releases-namespace/releases-project/-/archive/v1.1/releases-project-v1.1.tar.bz2",
          },
          Object {
            "format": "tar",
            "url": "http://localhost/releases-namespace/releases-project/-/archive/v1.1/releases-project-v1.1.tar",
          },
        ],
      },
      "author": Object {
        "avatarUrl": "https://www.gravatar.com/avatar/16f8e2050ce10180ca571c2eb19cfce2?s=80&d=identicon",
        "username": "administrator",
        "webUrl": "http://localhost/administrator",
      },
      "commit": Object {
        "shortId": "b83d6e39",
        "title": "Merge branch 'branch-merged' into 'master'",
      },
      "commitPath": "http://localhost/releases-namespace/releases-project/-/commit/b83d6e391c22777fca1ed3012fce84f633d7fed0",
      "descriptionHtml": "<p data-sourcepos=\\"1:1-1:33\\" dir=\\"auto\\">Best. Release. <strong>Ever.</strong> <gl-emoji title=\\"rocket\\" data-name=\\"rocket\\" data-unicode-version=\\"6.0\\">ðŸš€</gl-emoji></p>",
      "evidences": Array [
        Object {
          "collectedAt": "2018-12-03T00:00:00Z",
          "filepath": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/evidences/1.json",
          "sha": "760d6cdfb0879c3ffedec13af470e0f71cf52c6cde4d",
        },
      ],
      "milestones": Array [
        Object {
          "description": "The 12.4 milestone",
          "id": "gid://gitlab/Milestone/124",
          "issueStats": Object {
            "closed": 1,
            "total": 4,
          },
          "stats": undefined,
          "title": "12.4",
          "webPath": undefined,
          "webUrl": "/releases-namespace/releases-project/-/milestones/2",
        },
        Object {
          "description": "The 12.3 milestone",
          "id": "gid://gitlab/Milestone/123",
          "issueStats": Object {
            "closed": 3,
            "total": 5,
          },
          "stats": undefined,
          "title": "12.3",
          "webPath": undefined,
          "webUrl": "/releases-namespace/releases-project/-/milestones/1",
        },
      ],
      "name": "The first release",
      "releasedAt": "2018-12-10T00:00:00Z",
      "tagName": "v1.1",
      "tagPath": "/releases-namespace/releases-project/-/tags/v1.1",
      "upcomingRelease": true,
    },
  ],
  "paginationInfo": Object {
    "endCursor": "eyJpZCI6IjEifQ",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "eyJpZCI6IjEifQ",
  },
}
`;

exports[`releases/util.js convertOneReleaseGraphQLResponse matches snapshot 1`] = `
Object {
  "data": Object {
    "_links": Object {
      "closedIssuesUrl": "http://localhost/releases-namespace/releases-project/-/issues?release_tag=v1.1&scope=all&state=closed",
      "closedMergeRequestsUrl": "http://localhost/releases-namespace/releases-project/-/merge_requests?release_tag=v1.1&scope=all&state=closed",
      "editUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/edit",
      "mergedMergeRequestsUrl": "http://localhost/releases-namespace/releases-project/-/merge_requests?release_tag=v1.1&scope=all&state=merged",
      "openedIssuesUrl": "http://localhost/releases-namespace/releases-project/-/issues?release_tag=v1.1&scope=all&state=opened",
      "openedMergeRequestsUrl": "http://localhost/releases-namespace/releases-project/-/merge_requests?release_tag=v1.1&scope=all&state=opened",
      "self": "http://localhost/releases-namespace/releases-project/-/releases/v1.1",
      "selfUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1",
    },
    "assets": Object {
      "count": 8,
      "links": Array [
        Object {
          "directAssetUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/downloads/binaries/awesome-app-3",
          "external": true,
          "id": "gid://gitlab/Releases::Link/13",
          "linkType": "image",
          "name": "Image",
          "url": "https://example.com/image",
        },
        Object {
          "directAssetUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/downloads/binaries/awesome-app-2",
          "external": true,
          "id": "gid://gitlab/Releases::Link/12",
          "linkType": "package",
          "name": "Package",
          "url": "https://example.com/package",
        },
        Object {
          "directAssetUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/downloads/binaries/awesome-app-1",
          "external": false,
          "id": "gid://gitlab/Releases::Link/11",
          "linkType": "runbook",
          "name": "Runbook",
          "url": "http://localhost/releases-namespace/releases-project/runbook",
        },
        Object {
          "directAssetUrl": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/downloads/binaries/linux-amd64",
          "external": true,
          "id": "gid://gitlab/Releases::Link/10",
          "linkType": "other",
          "name": "linux-amd64 binaries",
          "url": "https://downloads.example.com/bin/gitlab-linux-amd64",
        },
      ],
      "sources": Array [
        Object {
          "format": "zip",
          "url": "http://localhost/releases-namespace/releases-project/-/archive/v1.1/releases-project-v1.1.zip",
        },
        Object {
          "format": "tar.gz",
          "url": "http://localhost/releases-namespace/releases-project/-/archive/v1.1/releases-project-v1.1.tar.gz",
        },
        Object {
          "format": "tar.bz2",
          "url": "http://localhost/releases-namespace/releases-project/-/archive/v1.1/releases-project-v1.1.tar.bz2",
        },
        Object {
          "format": "tar",
          "url": "http://localhost/releases-namespace/releases-project/-/archive/v1.1/releases-project-v1.1.tar",
        },
      ],
    },
    "author": Object {
      "avatarUrl": "https://www.gravatar.com/avatar/16f8e2050ce10180ca571c2eb19cfce2?s=80&d=identicon",
      "username": "administrator",
      "webUrl": "http://localhost/administrator",
    },
    "commit": Object {
      "shortId": "b83d6e39",
      "title": "Merge branch 'branch-merged' into 'master'",
    },
    "commitPath": "http://localhost/releases-namespace/releases-project/-/commit/b83d6e391c22777fca1ed3012fce84f633d7fed0",
    "descriptionHtml": "<p data-sourcepos=\\"1:1-1:33\\" dir=\\"auto\\">Best. Release. <strong>Ever.</strong> <gl-emoji title=\\"rocket\\" data-name=\\"rocket\\" data-unicode-version=\\"6.0\\">ðŸš€</gl-emoji></p>",
    "evidences": Array [
      Object {
        "collectedAt": "2018-12-03T00:00:00Z",
        "filepath": "http://localhost/releases-namespace/releases-project/-/releases/v1.1/evidences/1.json",
        "sha": "760d6cdfb0879c3ffedec13af470e0f71cf52c6cde4d",
      },
    ],
    "milestones": Array [
      Object {
        "description": "The 12.4 milestone",
        "id": "gid://gitlab/Milestone/124",
        "issueStats": Object {
          "closed": 1,
          "total": 4,
        },
        "stats": undefined,
        "title": "12.4",
        "webPath": undefined,
        "webUrl": "/releases-namespace/releases-project/-/milestones/2",
      },
      Object {
        "description": "The 12.3 milestone",
        "id": "gid://gitlab/Milestone/123",
        "issueStats": Object {
          "closed": 3,
          "total": 5,
        },
        "stats": undefined,
        "title": "12.3",
        "webPath": undefined,
        "webUrl": "/releases-namespace/releases-project/-/milestones/1",
      },
    ],
    "name": "The first release",
    "releasedAt": "2018-12-10T00:00:00Z",
    "tagName": "v1.1",
    "tagPath": "/releases-namespace/releases-project/-/tags/v1.1",
    "upcomingRelease": true,
  },
}
`;
